src/prng/isaac.rs

```diff
@@ -115,62 +116,21 @@ impl fmt::Debug for IsaacRng {
     }
 }
 
-fn mix(a: &mut w32, b: &mut w32, c: &mut w32, d: &mut w32,
-       e: &mut w32, f: &mut w32, g: &mut w32, h: &mut w32) {
-    *a ^= *b << 11; *d += *a; *b += *c;
-    *b ^= *c >> 2;  *e += *b; *c += *d;
-    *c ^= *d << 8;  *f += *c; *d += *e;
-    *d ^= *e >> 16; *g += *d; *e += *f;
-    *e ^= *f << 10; *h += *e; *f += *g;
-    *f ^= *g >> 4;  *a += *f; *g += *h;
-    *g ^= *h << 8;  *b += *g; *h += *a;
-    *h ^= *a >> 9;  *c += *h; *a += *b;
-}
-
 impl IsaacRng {
     /// Creates an ISAAC random number generator using an u64 as seed.
     /// If `seed == 0` this will produce the same stream of random numbers as
     /// the reference implementation when used unseeded.
     pub fn new_from_u64(seed: u64) -> IsaacRng {
-        let mut a = w(0x1367df5a);
-        let mut b = w(0x95d90059);
-        let mut c = w(0xc3163e4b);
-        let mut d = w(0x0f421ad8);
-        let mut e = w(0xd92a4a78);
-        let mut f = w(0xa51a3c49);
-        let mut g = w(0xc4efea1b);
-        let mut h = w(0x30609119);
-
-        let mut mem = [w(0); RAND_SIZE];
-
-        a += w(seed as u32);
-        b += w((seed >> 32) as u32);
-
-        for i in (0..RAND_SIZE/8).map(|i| i * 8) {
-            mix(&mut a, &mut b, &mut c, &mut d, &mut e, &mut f, &mut g, &mut h);
-            mem[i  ] = a; mem[i+1] = b;
-            mem[i+2] = c; mem[i+3] = d;
-            mem[i+4] = e; mem[i+5] = f;
-            mem[i+6] = g; mem[i+7] = h;
-        }
+        let mut key = [w(0); RAND_SIZE];
+        key[0] = w(seed as u32);
+        key[1] += w((seed >> 32) as u32);
+        // Initialize with only one pass.
         // A second pass does not improve the quality here, because all of
         // the seed was already available in the first round.
```

Comment:
I don't really get this — I mean, all of the seed gets used in the first round when using big keys too, right?

In reality I suppose it shouldn't matter since a 64-bit seed is not really sufficient for cryptography and this method should still be sufficient for non-crypto uses I guess (unless there's any detectable bias in initial output). Speaking very much as a non-expert here...

Reply:
For a short overview: in blocks of 8 words the seed is mixed with the temporary array `a`...`h`. Each block of 8 words then becomes part of the new state. After one full pass every word of the new state is effected by the first block. But only the last block of 8 words is effected by the last block of the seed.

After the first pass `a`...`h` has all of the seed mixed into it. Doing another pass will now mix bits from the middle and last blocks of the seed into the entire state.

The 64-bit seed is part of the first block, and therefore available to `a`...`h` from the first round. At the end of the first full pass the 'random quality' of `a`...`h` is not any higher than at the start. A second pass does not distribute the randomness from the seed more evenly than the first.

Reply:
Ah, I get you. But the 32-bit generator has two blocks set; `b` will not affect `a` until the second round if I understand correctly. I don't see why it matters though (provided `a` is initially set to a good seed, which it should be, since it gets added to the golden ratio).